### PR TITLE
UHM-6747, navigate to the question with invalid fields

### PR DESCRIPTION
**Pull request by cioan**
@mogoodrich , here is the PR that changes the current behavior of the component in such a way that when the user navigates ahead it nows stops at the in between question with invalid fields. This obvious now is in the contradiction of the unit test that says do not navigate ahead if the in between question has an invalid field:

https://github.com/openmrs/openmrs-module-uicommons/blob/master/omod/src/test/webapp/resources/scripts/navigatorHandlers.js#L715

### CODE REVIEW

**Review by mogoodrich — APPROVED** [review]
This looks good to me.  And I see what you mean now about this test:

"should not switch selection to question ahead if question in between is not valid"

Yes, I would rewrite that test to something like:

"should not switch selection to question ahead but should switch in invalid question in between"

Thanks Cosmin!

**Comment by cioan** [comment]
Thanks @mogoodrich ! I have updated the test cases to reflect the new navigation workflow. They are all green now. Thanks!

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js [hunk 1]
```diff
@@ -724,12 +725,15 @@ describe("Tests for simple form navigation handlers", function() {
...
                 field.isSelected = true;
                 firstQuestion.fields = [field];
+                spyOn(secondField, 'select');
+                spyOn(secondField, 'toggleSelection');
+                spyOn(secondField, 'isValid').and.returnValue(false);
+                secondQuestion.fields = [ secondField ];
 
                 clickedQuestionHandler(questions, thirdQuestion, event);
 
                 expect(firstQuestion.onExit).not.toHaveBeenCalled();
```
Comment (mogoodrich) [inline omod/src/test/webapp/resources/scripts/navigatorHandlers.js:735]: I guess I'm going down the rabbit hole here, but I *would* expect firstQuestion.onExit to have been called.
As Dave had mentioned, this whole path is a bit of an edge case, so maybe not worth spending all the time figuring out why this has't been called... perhaps change this to "expect(firstQuestion.onExit).toHaveBeenCalled(); but comment it out and add a TODO to fix in the future?

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js [hunk 2]
```diff
@@ -724,12 +725,15 @@ describe("Tests for simple form navigation handlers", function() {
...
                 expect(event.stopPropagation).toHaveBeenCalled();
-                expect(field.select).toHaveBeenCalled();
```
Comment (mogoodrich) [inline omod/src/test/webapp/resources/scripts/navigatorHandlers.js:729]: I would has guessed that the following would be the expected behavior in this case... do any of these lines pass? (ie, the use clicks on the third question, but the second it invalid, so the first question is toggled and existed, and the section question and field gets selected) 

`expect(event.stopPropagation).toHaveBeenCalled();
                expect(firstQuestion.onExit).toHaveBeenCalled();
                expect(firstQuestion.toggleSelection).toHaveBeenCalled();
                expect(secondQuestion.toggleSelection).toHaveBeenCalled();
expect(secondField.select).toHaveBeenCalled()`

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js [hunk 3]
```diff
@@ -784,7 +788,7 @@ describe("Tests for simple form navigation handlers", function() {
                 clickedQuestionHandler(questions, secondQuestion, event);
 
                 expect(event.stopPropagation).toHaveBeenCalled();
-                expect(field.select).toHaveBeenCalled();
```
Comment (mogoodrich) [inline omod/src/test/webapp/resources/scripts/navigatorHandlers.js:784]: I wouldn't expect this to be failed now, was it?

**Review by mogoodrich — CHANGES_REQUESTED** [review]
Sorry @cioan , maybe a rabbit hole :) but these test changes don't seem quite right, see my comments.

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js (hunk 3 above)
Comment (cioan) [reply omod/src/test/webapp/resources/scripts/navigatorHandlers.js]: I added it back.

**Comment by cioan** [comment]
Thanks @mogoodrich ! I take back everything bad I said about these test cases. In this case, actually it uncovered a potential bug with the onExit handler.

**Comment by mogoodrich** [comment]
Cool, thanks @cioan !  I won't have time to look at this until I'm back next week, but feel free to assign the ticket to me and I will check it out then.

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js [hunk 4]
```diff
@@ -712,21 +713,27 @@ describe("Tests for simple form navigation handlers", function() {
...
+                spyOn(secondField, 'select');
+                spyOn(secondField, 'toggleSelection');
+                spyOn(secondField, 'isValid').and.returnValue(false);
+                secondQuestion.fields = [ secondField ];
 
                 clickedQuestionHandler(questions, thirdQuestion, event);
 
-                expect(firstQuestion.onExit).not.toHaveBeenCalled();
+                expect(firstQuestion.onExit).toHaveBeenCalled();
                 expect(event.stopPropagation).toHaveBeenCalled();
```
Comment (mogoodrich) [inline omod/src/test/webapp/resources/scripts/navigatorHandlers.js:736]: Shouldn't the following new test cases be added?

```
expect(secondQuestion.toggleSelection).toHaveBeenCalled();
 expect(secondField.select).toHaveBeenCalled()
```

(Or at least the second of the two above?)

File: omod/src/test/webapp/resources/scripts/navigatorModels.js [hunk 5]
```diff
@@ -219,13 +219,14 @@ describe("Test for simple form models", function() {
...
             questionModel = new QuestionModel();
-            firstField = jasmine.createSpyObj('firstField', ['unselect', 'resetErrorMessages', 'value', 'displayValue', 'enable', 'disable', 'hide', 'show']);
+            firstField = jasmine.createSpyObj('firstField', ['unselect', 'resetErrorMessages', 'value', 'displayValue', 'enable', 'disable', 'hide', 'show', 'isValid']);
             secondField = jasmine.createSpyObj('secondField', ['unselect', 'resetErrorMessages', 'value', 'displayValue', 'enable', 'disable', 'hide', 'show']);
             firstField.element = jasmine.createSpyObj('element', ['addClass', 'removeClass', 'hasClass']);
             secondField.element  = jasmine.createSpyObj('element', ['addClass', 'removeClass', 'hasClass']);
             questionModel.fields = [firstField, secondField];
             element = jasmine.createSpyObj('element', ['addClass', 'removeClass', 'triggerHandler', 'attr', 'removeAttr', 'hide', 'show']);
             questionModel.element = element;
+            spyOn(questionModel, 'isValid').and.returnValue(true);
```
Comment (mogoodrich) [inline omod/src/test/webapp/resources/scripts/navigatorModels.js:229]: It seems odd to me that the test "should state the question is invalid" to pass with this line added... specifically, see line 392.  Maybe I'm missing something>

**Review by mogoodrich — APPROVED** [review]
Thanks @cioan , code looks good to me, thought I added a couple small questions about the tests....

File: omod/src/test/webapp/resources/scripts/navigatorHandlers.js (hunk 4 above)
Comment (cioan) [reply omod/src/test/webapp/resources/scripts/navigatorHandlers.js:738]: Thanks @mogoodrich ! I have added the following two validation statements:
`expect(secondQuestion.toggleSelection).toHaveBeenCalled();
                expect(secondField.toggleSelection).toHaveBeenCalled();`

If you look at the handler code, the field.toggleSelection() function is called:
https://github.com/openmrs/openmrs-module-uicommons/pull/107/files#diff-843c954a225b484145f9fd50f0ef831655d7001490c56a23c9c1053f4c4ad1cbR300

File: omod/src/test/webapp/resources/scripts/navigatorModels.js (hunk 5 above)
Comment (mogoodrich) [reply omod/src/test/webapp/resources/scripts/navigatorModels.js:229]: sorry, line 392 in the same file, I was referring to this test:

https://github.com/openmrs/openmrs-module-uicommons/blob/master/omod/src/test/webapp/resources/scripts/navigatorModels.js#L383

Specifically, on line 229, if I understand it correctly, we are mocking the questionModel 'isValid' method and telling it to return true, so I would assume line 392 would set isValid=true, yet 396 expects it to be false... so I would have expected the test starting on line 383 to be failing, but if it isn't I'm likely missing something.

**Review by mogoodrich — COMMENTED** [review]
@cioan added a comment that hopefully clarifies my previous comment re: line 392

**Comment by mogoodrich** [comment]
Yes, I realize you hadn't touched that test.  I also think I see what I had been missing...

You are correct, setting "firstField.isValid.and.returnValue(false)" should make the questionModel.isValid() method return "false", but in the beforeEach we now set "spyOn(questionModel, 'isValid').and.returnValue(true)" I think this basically says "when calling questionModel.isValid() don't actually call the isValid method, just return true in all cases".  So I was surprised why this test didn't start to fail with the addition of line 229.

But I see now that that test doesn't use the questionModel set up in beforeEach, but rather initializes a new questonModel:
https://github.com/openmrs/openmrs-module-uicommons/blob/master/omod/src/test/webapp/resources/scripts/navigatorModels.js#L389

So I think that's why the test doesn't break.

Anyway, the tl;dr I should have added earlier... lgtm, go ahead merge!  Thanks!

**Pull request merged**